### DISCUSSION TITLE
More Performance Improvements

### DISCUSSION THREAD

**Discussion by engsky** [oqtane/oqtane.framework#2487] [Q&A] [answered]
Replacing AddDbContext with AddDbContextPool. What important points should be taken in care? for example one of these points is  varying tenant.

**Comment by sbwalker** [comment, marked as answer]
I have tried to use the DbContextPool on other projects and I have not had much success. The main challenge is the service scope - DbContextPool is essentially a singleton which means it can only be used in scenarios where the configuration is completely static (ie. the DbContext is configured once for the life of the application and never changes). Very few enterprise applications fit this scenario as they typically have some degree of variability in configuration (ie. they implement OnConfiguring() to override specific properties for each instance). In particular, multi-tenant applications like Oqtane need to override the connection string to point at the appropriate tenant database based on run-time characteristics. Therefore, DbContextPool is not a good fit for these types of applications. I would also say that although there is documentation which suggests there are some theoretical performance benefits of using DbContextPool, this same documentation fails to mention the strict limitations or challenges of using this feature (which is typical as most developer documentation is actually marketing propaganda). I have never seen a real world performance benchmark of this feature but most engineers I have spoken to have said the improvements are minimal at best. This is a good thread to read on this topic - https://github.com/dotnet/efcore/issues/14169 - and notice the comments from the Engineering Manager for .NET Data and Entity Framework at Microsoft.
